### DISCUSSION TITLE
BUG: Router decorators throw error converting to `response_model`  when alias is present in returned model instance

### DISCUSSION THREAD

**Discussion by tomyo** [fastapi/fastapi#9166] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter
from pydantic import BaseModel, Field


router: APIRouter = APIRouter(
    prefix="/bug",
    tags=["bug"],
)


class ModelInDB(BaseModel):
    eid: str = Field(..., alias="_id")


class ModelResponse(BaseModel):
    eid: str

    class Config:
        orm_mode = True


@router.get("/", response_model=ModelResponse)
async def get_scheduled_goals() -> ModelInDB:
    return ModelInDB(_id="test")
```

### Description

Call GET endpoint `/bug`.
A 500 server error occurs with trace:

```sh
INFO:     127.0.0.1:53128 - "GET /v1/bug/ HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/base.py", line 25, in __call__
    response = await self.dispatch_func(request, self.call_next)
  File "~/projects/macrosfirst/integrations/./app/main.py", line 64, in db_session_middleware
    response: Response = await call_next(request)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/base.py", line 45, in call_next
    task.result()
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/base.py", line 38, in coro
    await self.app(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/middleware/cors.py", line 78, in __call__
    await self.app(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/fastapi/routing.py", line 224, in app
    response_data = await serialize_response(
  File "~/.pyenv/versions/3.8.15/lib/python3.8/site-packages/fastapi/routing.py", line 127, in serialize_response
    raise ValidationError(errors, field.type_)
pydantic.error_wrappers.ValidationError: 1 validation error for ModelResponse
response -> eid
  field required (type=value_error.missing)
```

### Operating System

Linux

### Operating System Details

5.15.93-1-MANJARO

### FastAPI Version

tested on 0.65.3 and 0.69.0 

### Python Version

3.8.15

### Additional Context

If I stop the debugger at the end of the router function, I can successfully run:

```python

instance = ModelInDB(_id="test")
expected_result = ModelResponse.from_orm(instance)
should_be = ModelResponse(eid='test')

expected_result == should_be  # True
```

That means FastAPI is not able to automatically perform this conversion.

Looking at the source code, I see it always tries to do `res.dict(by_alias=True)` in `prepare_response_content`, when `isinstance(res, BaseModel)`, which is the case.

This looks like a bug to me.

Current workaround is to return a dict with the expected fields instead of returning the model instance, but loosing typing hints:

```python
@router.get("/", response_model=ModelResponse)
async def get_scheduled_goals() -> dict:
    return ModelInDB(_id="test").dict()
```
    
Additional Note / BUG about 0.69 version:

Version `0.69` even has an additional bug in `prepare_response_content` function, where it checks for  `read_with_orm_mode`  instead of `orm_mode`.

`fastapi/routing.py:72`
```python
read_with_orm_mode = getattr(res.__config__, "read_with_orm_mode", None)
        if read_with_orm_mode:
            # Let from_orm extract the data from this model instead of converting
            # it now to a dict.
            # Otherwise there's no way to extract lazy data that requires attribute
            # access instead of dict iteration, e.g. lazy relationships.
            return res
```

If I include that field name in ModelResponse's config class, it works:
```python

class Config:
    read_with_orm_mode=True
```

But IDE shows a type error as that field is not declared as a valid one to begin with.

Thank you for looking into this. Happy to help in a PR/reporting the issue if needed.

P.S, I see in docs no way to specify if `by_alias` is used for converting the returned object to the declared `model_response` in the router. I think this would be a nice addition, if it can't be done automatically.

**Comment by ebottos94** [comment]
Hi @tomyo, your error is in response_model, ModelResponse has no reference to _id, you have two choices : 

1) 

```python
from fastapi import APIRouter
from pydantic import BaseModel, Field


router: APIRouter = APIRouter(
    prefix="/bug",
    tags=["bug"],
)


class ModelInDB(BaseModel):
    eid: str = Field(..., alias="_id")


class ModelResponse(BaseModel):
    eid: str

    class Config:
        orm_mode = True


@router.get("/", response_model=ModelInDB)
async def get_scheduled_goals() -> ModelInDB:
    return ModelInDB(_id="test")
```

2)

```python
from fastapi import APIRouter
from pydantic import BaseModel, Field


router: APIRouter = APIRouter(
    prefix="/bug",
    tags=["bug"],
)


class ModelInDB(BaseModel):
    eid: str = Field(..., alias="_id")


class ModelResponse(ModelInDB):

    class Config:
        orm_mode = True


@router.get("/", response_model=ModelResponse)
async def get_scheduled_goals() -> ModelInDB:
    return ModelInDB(_id="test")
```

**Reply by ebottos94** [reply]
@tomyo if your problem is solved can you mark the comment as the answer ? Thank you.

**Reply by tomyo** [reply]
Thanks for your reply, @ebottos94. 

Both models have the same field name (`eid`), that's why I think this is a bug, or at least there should be a way to make this work as is.

About your suggestions: I think you're missing the point of discussing a possible bug report, and suggesting ways to work around the issue instead.
Still, I'll reply on the suggestions:

* On point 1:
  I can't use `ModelInDB` in `response_model`, since then it will show the field `_id` (the alias) for the schema in /docs instead of the field name.  Why there's no way to specify where should the docs look for (field name vs alias) for building the schemas, is beyond my comprehension, but you are right in the point that `ModelReponse` could be omitted if fastAPI would allow the flexibility to specify how to build the schemas in /docs. The only reason I have that model, is to make /docs show the right schema.  So, this suggestion breaks the API schema I want to expose.

* On point 2:
  It suffers from the same issue described above: Docs will say the schema has an `_id` field, when what I want is to work with the field `eid` in the api, as `_id` is the DB field name instead.
  
 
Is there something else you think I'm missing here?  Thanks again!

**Reply by ebottos94** [reply]
ok @tomyo, now I understood your issue. 

You can use this approach :  

```python
from fastapi import APIRouter
from pydantic import BaseModel, Field


router: APIRouter = APIRouter(
    prefix="/bug",
    tags=["bug"],
)


class ModelInDB(BaseModel):
    eid: str = Field(alias="_id")



class ModelResponse(BaseModel):
    eid : str

    class Config:
        orm_mode = True


@router.get("/", response_model=ModelResponse)
async def get_scheduled_goals() -> ModelResponse:
    return ModelInDB(_id="test").dict(by_alias=False)
```

**Reply by andruli** [reply]
Isn't this just a workaround?

I know the code you suggested works, but it its counterintuitive to me. **I'd expect fastapi to behave just like pydantic** with the `from_orm_mode` but it doesn't.

**Comment by YuriiMotov** [comment, marked as answer]
Works fine in current version
